# Farshid Pirahansiah
# Config jupyter lab for computer vision 

In [3]:
#!pip install line_profiler
#conda install -c conda-forge ipympl
%matplotlib widget
#%matplotlib inline
#%matplotlib notebook
#%load_ext line_profiler
#%run
#%pdb #Automatically enters the Python debugger after any uncaught exception. A great tool for debugging.
#%prun #Runs code with the Python profiler to identify which internal parts of the function are taking the longest to execute.
#%%capture



In [1]:
text="""
"
<span style="color: red;">2024.</span>
<span style="color: green;">Edge-based Generative AI for Image Processing Aplication</span>
On-device processing: Prioritizing privacy and accessibility, LLMs on-device processing, ensuring that your data remains on your device. This not only enhances security but also has the potential to minimize latency, contributing to a more efficient user experience.

Generative AI models are a class of AI that can generate new data that resembles existing data. They include foundation models that can be fine-tuned for different tasks, Variational Autoencoders (VAEs) that reduce the dimensionality of data and Generative Adversarial Networks (GANs) that use competing networks to generate realistic samples. Other models include Transformer-based models that use attention mechanisms to model long-term text dependencies, diffusion models that address information decay by removing noise in the latent space, and multimodal large language models (LLMs) that integrate image and text understanding.
In summary, generative AI models are a class of AI that can generate new data that resembles existing data. They include foundation models, VAEs, GANs, Transformer-based models, diffusion models, and multimodal LLMs.

# Farshid PirahanSiah
## Generative AI models on Edge with on-device training 
- CNN, RNN, Transformer-based models, LLMs (GPT, )
- Generative AI models
    - foundation model
        - train on huge data -> adapt to applications
        - GPT, CLIP, DALL-E, 
    - Variational Autoencoders (VAEs)
        - rapidly reduce the dimensionality of samples
        - input (image) -> encoder -> latent space -> decoder -> reconstructed output (image)
        - use for 
            - image synthesis
            - data compression
            - anomaly detection
        - **Standard VAEs**
            - **Beta-VAE**: Balances latent capacity and reconstruction.
        - **Conditional VAEs (CVAEs)**
            - **Conditional Beta-VAE**: Adds conditional variables for targeted generation.
        - **Hierarchical VAEs**
            - **NVAE**: Deep hierarchy, residual connections for stable high-quality output.
    - Generative Adversarial Networks (GANs)
        - use competing networks to produce realistic samples
        - generator ; discriminator
        - GANs in finance, spaceGAN (geospatial data), styleGAN2 (create video game characters)
        - **Convolutional GANs**
            - **DCGAN**: Deep convolutional networks for stable, quality generation.
        - **Style-based GANs**
            - **StyleGAN**: Fine-grained style control, realistic faces.
            - **StyleGAN2**: Redesigns normalization, reduces artifacts.
        - **Progressive Growing GANs**
            - **PGGAN**: Increases network size during training for high-resolution output.
    - Transformer-based Models
        - use attention mechanisms to model long-term text dependencies
        - **Autoregressive Models**
            - **GPT-3**: Predictive text generation based on input.
            - **Image GPT (iGPT)**: Generates images pixel by pixel.
        - **Multimodal Transformers**
            - **DALL-E**: Generates images from text descriptions.
    - Diffusion Models
        - address information decay by removing noise in the latent space
        - step 1: forward diffusion to add random noise to the data; 
        - step 2: reverse diffusion: turn the noise , recover data, generate the desired output
        - **Basic Diffusion Models**
            - **DDPM**: Generates data from noise by reversing diffusion.
        - **Advanced Diffusion Techniques**
            - **Improved DDPM**: More efficient, improved sampling.
            - **Guided Diffusion**: Produces specific outputs with classifier guidance.
    - Multimodal Large Language Models (LLMs)
        - **CLIP**
            - Integrates image and text understanding.
        - **Perceiver IO**
            - Handles audio, visual, text with a generalized architecture.
        - **FLAVA**
            - Self-supervised, unified model for vision, language, and multimodal tasks.
- LLMOps: MLOps Tools: MLflow and Hugging Face    
    - <span style="color: green; font-weight: bold; font-style: italic; white-space: nowrap;"> 📚 🤔 MLflow <span style="display: inline-block; animation: slide 2s infinite linear;">→</span> </span>  <style>            @keyframes slide {            0% { transform: translateX(0); }            100% { transform:translateX(20px); }            }            </style>
 
        - MLflow Tracking - Logs key metrics, parameters, models, and other artifacts when running ML code to monitor experiments
        - MLflow Projects - Configurable standard format for organizing ML code to ensure consistency and reproducibility
        - MLflow Models - Package ML model files with their dependencies so they can be deployed on diverse platforms
        - pip install mlflow
        - mlflow ui
        - mlflow experiments create --experiment-name metrics-test ==> id 2
        - MLFLOW_EXPERIMENT_ID=2 python test-mlflow.py
        - load_model= mlflow.pyfunc.load_model(logged_model)
        - mlflow run . -P filename=inputfile
        - hugging face 
            - models, dataset, spaces
            - streamlit, gradio
            - pip install huggingface_hub
            - huggingface-cli login
            - git lfs install
            - git clone <model url>
            - 
        - 









## Large Vision Models (LVMs)
-  Stable Diffusion
    - Why Stable Diffusion 
        - **Brainstorming and Ideation:** Rapidly generate images to explore various design layouts, scenes, or concepts. It’s an efficient tool for experimenting with creative ideas, such as different settings or stylized visual effects, providing more specific and customized placeholders than stock photos or traditional sketches.
        - **Character Design:** Whether for video games, animations, or comic books, Stable Diffusion can help conceptualize and visualize characters, aiding in the development and refinement of unique designs.
        - **Storyboarding:** Create visual narratives for video productions, movies, or advertisements. Stable Diffusion can generate scenes and sequences, facilitating the storytelling process and the visualization of cinematic ideas.
        - **Communication:** Enhance presentations or proposals with high-quality visualizations. For office managers, interior designers, or event planners, it’s a tool to convey ideas about space layouts, mood, atmosphere, and color schemes effectively, even if the images aren’t perfect replicas.
        - **Asset Generation:** Produce elements for compositing into photographs or artworks, create textures for 3D modeling, or develop backgrounds for motion graphics. It’s a versatile tool for augmenting traditional and digital art processes with unique visual components.
        - **Fun and Exploration:** Beyond professional applications, Stable Diffusion offers a platform for fun and creative exploration. It encourages playful experimentation with images and concepts, often leading to unexpected and delightful results.  
    - **Stable Diffusion**
        - **Image Style**
            - Blends photorealism and illustration
            - Customizable styles via data models
        - **Data Models**
            - Open-source, community-driven enhancements
            - Customizable for content safety or style
        - **Accessibility**
            - Free and open-source
            - Standalone applications
            - Integrates with software like Photoshop
        - **Ethics**
            - Training on diverse datasets
            - Ethical considerations similar to competitors
        - **Features**
            - Image-to-image generation
            - Custom training for specific recognition
        - **Compared to Others**
            - DALL-E and Midjourney focus on photorealism
            - Adobe Firefly trained on ethical datasets
            - Midjourney accessible via Discord
            - Others require payment, Stable Diffusion is free
        - **Community Impact**
            - Vibrant open-source community
            - Continuous improvement and feature additions
        - **Technology**
            - AI image generator revolution
            - Different from traditional digital tools
            - Democratizes sophisticated image generation






"""


In [2]:
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM, pipeline

model_name = "allenai/led-large-16384-arxiv"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSeq2SeqLM.from_pretrained(model_name)

# If you are running on a GPU, you can uncomment the device parameter to utilize it
# pipe = pipeline("text2text-generation", model=model, tokenizer=tokenizer, device=0)
pipe = pipeline("text2text-generation", model=model, tokenizer=tokenizer)

long_text = text
generated_text = pipe(
    long_text, 
    truncation=True, 
    max_length=512,  # Adjusted to a higher value for potentially longer outputs
    no_repeat_ngram_size=5, 
    num_beams=3, 
    early_stopping=True
)

# Printing the generated text to see the output
print(generated_text)

Input ids are automatically padded from 2756 to 3072 to be a multiple of `config.attention_window`: 1024


[{'generated_text': ' generative models are a class of artificial intelligence ( AI ) that can generate new data that resembles existing data. \n they include foundation models that can be fine-tuned for different tasks , Variational Autoencoders (VAEs ) that reduce the dimensionality of data and Generative Adversarial Networks (GANs ) that use competing networks to generate realistic samples . \n other models include Transformer-based models that use attention mechanisms to model long-term text dependencies , diffusion models that address information decay by removing noise in the latent space , and multimodal large language models (LLMs ) that integrate image and text understanding.    \n generative models on Edge with on-device training - CNN, RNN , Transformer-based model , LLMs (GPT, )- Generative models on Edge  - foundation model , on-device training , on-device processing  - image synthesis , image compression , anomaly detection , data compression     * keywords : * generative

Input ids are automatically padded from 2756 to 3072 to be a multiple of `config.attention_window`: 1024
[{'generated_text': ' generative models are a class of artificial intelligence ( AI ) that can generate new data that resembles existing data. \n they include foundation models that can be fine-tuned for different tasks , Variational Autoencoders (VAEs ) that reduce the dimensionality of data and Generative Adversarial Networks (GANs ) that use competing networks to generate realistic samples . \n other models include Transformer-based models that use attention mechanisms to model long-term text dependencies , diffusion models that address information decay by removing noise in the latent space , and multimodal large language models (LLMs ) that integrate image and text understanding.    \n generative models on Edge with on-device training - CNN, RNN , Transformer-based model , LLMs (GPT, )- Generative models on Edge  - foundation model , on-device training , on-device processing  - image synthesis , image compression , anomaly detection , data compression     * keywords : * generative models , image processing , image synthesis , image synthesis , data compression , data compression , anomaly detection '}]